### PR TITLE
it's a start

### PR DESCRIPTION
**Pull request by peektism**
Rather than write the whole thing and get review once it's "done", I think I should get review on the existing code and then I can adjust to keep building.

- `node_id` is most complete and has a few tests, but is still missing a few implementations that will depend on how I implement stuff in `config` for the `NodeHandler` 
- `node` creates a `NodeInfo` from `NodeId`, `SocketAddr`, and `NodeStatus`
- `store` is a WIP -- I keep looking at lp2p and considering what to use and not use from there...it's good code, but it is maybe unnecessarily generic for our purposes
- `error` uses failure with the simplest syntax and `util` contains some test scaffolding that I expect to be useful for `config` and `store`
- `config` and `store` are the least complete, but they should pull everything together and actually make the whole thing usable/working

> should squash these useless commits if possible at some point

### CODE REVIEW

File: .gitignore
Comment (dvc94ch) [inline .gitignore]: Why are you ignoring the `Cargo.lock`?

File: .gitignore
Comment (dvc94ch) [inline .gitignore]: these shouldn't be in the `.gitignore`

File: Cargo.toml
Comment (dvc94ch) [inline Cargo.toml]: failure_derive is reexported in `failure`
bs58 is slow since it's not byte aligned. base64 or base32 are much much faster, but can leave for now

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: Why not use `Instant::now()`? It's guaranteed to be monotonic

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: Not sure this is a good idea, since the error can't be handled in any other way than retrying, which means we can just as well not have a timeout?

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: ret_val?

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: Why is this needed?

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs:72]: probably should check the difficulty?

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: this should go into the module where the tests are

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: todo

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs:141]: How about renaming metric_distance to distance and impl From<Metric> for NodeId. You probably need that config trait we talked about:
```rust
pub trait Dht {
    type NodeId: From<Self::Metric>; 
    type Metric;
}
```

File: src/store.rs
Comment (dvc94ch) [inline src/store.rs]: can you run rustfmt?

File: spec.md
Comment (dvc94ch) [inline spec.md:76]: I'm really not sure what bitswap has to do with this

File: spec.md
Comment (dvc94ch) [inline spec.md:10]: that would require a consensus mechansim, which is out of scope for the dht. you could provide a function to set the difficulty from a higher layer.

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: what is the purpose of this test? Are you checking that the hash function works correctly?

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: `let big_data = vec![1u8; 33];`

**Review by dvc94ch — CHANGES_REQUESTED** [review]
Looking good!

File: .gitignore
Comment (libreamartists) [reply .gitignore]: because it's so large, I can not ignore it as well

File: .gitignore
Comment (libreamartists) [reply .gitignore]: I assume you meant the `**/*.rs.bk` and not the empty line, just changed and will push soon

File: Cargo.toml
Comment (libreamartists) [reply Cargo.toml]: ok, will make an issue once this is merged

File: .gitignore
Comment (dvc94ch) [reply .gitignore]: I meant examples benches and your private files

File: Cargo.toml
Comment (dvc94ch) [reply Cargo.toml]: Depends how often you have to encode/decode. For cid's it's  a lot. For nodeids maybe not regularly.

File: spec.md
Comment (libreamartists) [reply spec.md:10]: It's not out of scope for s-kademlia, which theoretically requires a provably difficult and temper-proof mechanism to limit NodeId generation in a given time period. If we're just hashing public keys, then I can have a few machines doing that all the time to generate a set of public keys for which the respective hash maintains the requisite number of trailing zeros. Then, I can just sybil attack the network with the **pre-generated set of public keys** whenever I want. 

The current design is dumb -- it's not robust against sybil attacks and I don't want to pretend it is. Let's assume that it is provably difficult via benchmarking or hardware testing. This just means that normal nodes won't be able to generate a new `NodeId` in a reasonable amount of time and, if the puzzle is hard enough, it will restrict the network to the nodes that have the processing power to generate sets of keys and whoever they want to share their pre-generated keys with.

See the next bullet in this `spec` page, **I'm leaning towards using the blockchain to limit the number of nodes joining. The existing members can sponsor a quota number of new members every `JoinPeriod: BlockNumber`.**

Comment (libreamartists) [reply spec.md:10]: In the short to medium term, I could organize the code such that a new `NodeHandler` generates a `NodeId` with a `difficulty` input set by some `Protocol` in the `impl NodeHandler` `new` method. This makes a lot of sense and I'm thinking about it.

Comment (dvc94ch) [reply spec.md:10]: As I said you can provide an API to set the difficulty. But then don't forget that if you change the difficulty dynamically, everyone needs to generate a new identity

File: spec.md
Comment (libreamartists) [reply spec.md:76]: It's a data transfer protocol. I want to add data stores to the DHT and configure protocols to gossip the data to provide certain availability/consistency guarantees. I'd use bitswap to organize these gossip protocols. Were you thinking of putting this logic somewhere else like a `consensus` crate? I think our consensus crates should implement some `Protocol` trait that specifies some enforced bitswap strategy for each data store. Eventually, we can use proof of the enforcement/participation of bitswap strategies to manage the node membership set (on the blockchain).

File: .gitignore
Comment (libreamartists) [reply .gitignore]: why shouldn't it be in the `.gitignore`? I have them in the same directory and I don't want to commit them. I could create an `ignore` folder, put everything in there, and then put `ignore/` in `.gitignore` but I don't see significant value add. I guess I reveal some info about my planning process now, but I don't think that's a big deal.

File: spec.md
Comment (libreamartists) [reply spec.md:10]: why? I can change the difficulty and still accept old identities. I just need to come to consensus on the set of identities (as discussed three comments above).

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: That's fine, I really dislike this method anyway. It's a placeholder for what I am thinking will be a more robust anti-sybil mechanism that uses blockchain consensus to manage the set of identities.

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: If we don't have a timeout, I don't even need it. That's what I'm going to do, but this method is definitely just a placeholder...

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: I have no clue why I had this and the `counter`. Took them out now.

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs:72]: the difficulty is not a field of `NodeId`, do you think it should be?

File: src/store.rs
Comment (libreamartists) [reply src/store.rs]: done, will keep doing

File: .gitignore
Comment (dvc94ch) [reply .gitignore]: You have multiple options to handle this:
- don't stage the files
- commit them to a separate branch
- put them in a different folder

It's not what the .gitignore is for and in general, no private information should be in a git repo

File: src/node_id.rs
Comment (dvc94ch) [reply src/node_id.rs]: You need a dht to build a blockchain. It's a chicken and egg problem. Think of this as being a building block for building a blockchain, not a blockchain.

File: src/node_id.rs
Comment (dvc94ch) [reply src/node_id.rs:72]: No, but if you don't check the difficulty there's no point in even having it. Where do you propose checking it?

File: spec.md
Comment (dvc94ch) [reply spec.md:76]: Consensus has absolutely nothing to do with bitswap. Can you explain how bitswap helps building a dht?

File: spec.md
Comment (dvc94ch) [reply spec.md:10]: Chicken and egg. How do you plan to reach consensus without a dht?

Comment (dvc94ch) [reply spec.md:10]: Can we please not mix consensus, bitswap and dht? They're completely different things and should be treated as such.

File: .gitignore
Comment (dvc94ch) [reply .gitignore]: There's a reason why it's not in the gitignore by default

File: .gitignore
Comment (dvc94ch) [reply .gitignore]: https://stackoverflow.com/questions/5724455/can-i-make-a-user-specific-gitignore-file

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: You don't need to use this DHT to build a blockchain. You can use a blockchain as a primitive. It's not a chicken or egg problem. Just because every blockchain has a DHT doesn't mean that a DHT can't use a blockchain for managing the set of nodes. It doesn't have to be (and probably shouldn't be) a blockchain that uses this DHT.

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: true, this is useless and I moved the `random` to `util::test`

File: spec.md
Comment (libreamartists) [reply spec.md:76]: Consensus will define how we decide shared state with respect to data with some *consistency* requirements. Bitswap defines how the data is gossiped. The DHT's primary function will be discerning the set of nodes, but each node will also store data. The protocol (strategy) for requesting and sending data is defined by bitswap. Consensus helps us decide on shared state based on messages sent/received between nodes.

**I'm going to start with something simple first**, but I don't want to restrict the scope when I can see how blockchain would be useful for managing the set of nodes and coming to consensus on data. It will also be useful for proving availability, but that's a bit later...

File: spec.md
Comment (libreamartists) [reply spec.md:10]: Don't need to use the same DHT for consensus on the set of nodes.

**You said please so I'll drop this for now**

File: src/node_id.rs
Comment (dvc94ch) [reply src/node_id.rs]: so you're suggesting to use substrate to build this DHT?

File: .gitignore
Comment (libreamartists) [reply .gitignore]: unignored

File: spec.md
Comment (dvc94ch) [reply spec.md:76]: both bitswap and the dht deal with self authenticating data. the difference is bitswap cares about the hash of the data and the dht cares about the signature of the data. so one authenticates the data itself while the other authenticates the author.

Comment (dvc94ch) [reply spec.md:76]: and I'm not sure what bitswap has to do with gossip

File: .gitignore
Comment (libreamartists) [reply .gitignore]: I removed some of them, but `benches` isn't done or properly started.

File: spec.md
Comment (libreamartists) [reply spec.md:76]: bitswap defines how nodes request/send data -- I'm unsure how this decision can have nothing to do with gossip.

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs:72]: I don't. I don't think this (leading zeros of `hash(pubkey)`) is how limited `NodeId` generation should be enforced.

File: .gitignore
Comment (dvc94ch) [reply .gitignore]: so if I create a bench, I'm like wtf, where's my bench? The only rules that should be in the . gitignore are rules that hold under *all* circumstances for *all* devs

File: spec.md
Comment (dvc94ch) [reply spec.md:76]: gossip means efficiently broadcasting data to multiple subscribers (1:n). Bitswap is a protocol for exchanging (a specific kind of) data (1:1).

File: src/node_id.rs
Comment (dvc94ch) [reply src/node_id.rs:72]: So how do you enforce it and where? Are you relying on participants to self-enforce it because they're honest?

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: I'm suggesting to use substrate to limit `NodeId` generation for this DHT.

Comment (dvc94ch) [reply src/node_id.rs]: @troublescooter has some thoughts on how to build Sybil resistant dhts

File: src/node_id.rs
Comment (4meta5) [reply src/node_id.rs]: Will do tomorrow -- wait for this before merging...

File: src/node_id.rs
Comment (dvc94ch) [reply src/node_id.rs]: Can we open an issue for this? Depending on substrate is not something that we should decide without evaluating other options

File: .gitignore
Comment (libreamartists) [reply .gitignore]: ok, I'll fix this in the next push

File: spec.md
Comment (libreamartists) [reply spec.md:76]: ok, took out bitswap

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: should this be done before this is merged? Shouldn't be too hard, but I'd need to implement `sign_message` on `config::NodeHandler`

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: #9

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs:72]: taking this out and placing it in `benches/`

File: src/error.rs
Comment (dvc94ch) [inline src/error.rs]: `use failure::Fail;`

File: src/lib.rs
Comment (dvc94ch) [inline src/lib.rs]: You can import the macro directly in the 2018 edition with `use failure::Fail`.

File: src/node_id.rs
Comment (dvc94ch) [inline src/node_id.rs]: these don't have to be mut

File: src/store.rs
Comment (dvc94ch) [inline src/store.rs:11]: These should go into the `Dht` trait

**Review by dvc94ch — CHANGES_REQUESTED** [review]

File: src/error.rs
Comment (libreamartists) [reply src/error.rs]: good to know, changed now

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: true, I just changed it but it is just a test

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs]: was wrong about this, just implemented `verify` and added a test to sign and verify...

File: src/node_id.rs
Comment (libreamartists) [reply src/node_id.rs:141]: I'm actually confused about what you're looking for here? Should I include the `distance` method in this new trait as well?

```rust
pub trait Dht {
    type Err;
    type NodeId: From<Self::Metric>;
    type Metric;

    fn distance(&self, other: &Self) -> Result<Self::Metric, Self::Err>;
}
```

And then, I guess I would implement this for `config::NodeHandler`, but that doesn't really feel right. It seems you have something in mind and I can't guess what it is.

File: src/store.rs
Comment (libreamartists) [reply src/store.rs:11]: `Dht` trait confuses me. I made some changes to the `KadMetric` trait and think it's ready to merge.

**Pull request merged**
